### PR TITLE
Don't filter out non-GC memory pools in stats

### PR DESCRIPTION
**Pull request by rmuir**
Currently pools such as Code Cache are filtered out of stats... but this is throwing away potentially important diagnostic information.

There are many thousands of classes in this codebase, as well as dynamically generated classes (e.g. scripts). We also have to look out for bugs like https://bugs.openjdk.java.net/browse/JDK-8023191

The handling for an unbounded pool (-1) was wrong here. MemoryUsage.getUsed() cannot be negative. For getMax(), `This method returns -1 if the maximum memory size is undefined.` This is the case for e.g. Metaspace, so Long.MAX_VALUE makes a lot more sense than zero.

### CODE REVIEW

**Comment by nik9000** [comment]
LGTM. There is a test for this class that might be worth poking at as well.

**Comment by rmuir** [comment]
I think assigning tests to this would be fragile? These are system pools that are very implementation specific: even within one impl, they change within every release. In other words, just for diagnostics.

For example, with java 8, permgen changes to metaspace, and then under java 9, there are 3 codecache pools (http://openjdk.java.net/jeps/197) ...

So I think it is best to treat them as opaque.

**Comment by nik9000** [comment]
> I think assigning tests to this would be fragile?

I agree. The current tests are fairly unopinionated and ought to stay that way. I really brought it up in case you had any good ideas.

I don't have any *good* ideas, just bad ones: I don't think the Long.MAX_VALUE stuff is that important to test because 0 isn't the end of the world. Maybe pull the null handling into `GcNames.getByMemoryPoolName` and test that because it was the actual bug? It is a fairly lame thing to test but at least it is something.

I don't really want a test that says "I either saw metaspace or permgen and they look like this". That seems a needlessly brittle.

**Comment by rmuir** [comment]
Well the test already loops over all the pools. In my opinion, we should remove `GcNames` completely and not do any of this pool-"interpreting" we do already. 

For example, attempting to bucket them into "young" and "old" and so on makes no sense with some garbage collectors, with G1 you see some infinitely sized stuff, which is confusing, and with shenandoah GC you have none of this at all.

So the opaqueness of this patch helps in that case, with master today you have no information whatsoever, at least with this patch you will then see something:

```
            "Shenandoah" : {
              "used_in_bytes" : 177207040,
              "max_in_bytes" : 2147483648,
              "peak_used_in_bytes" : 208141344,
              "peak_max_in_bytes" : 2147483648
            }
```

**Comment by nik9000** [comment]
> In my opinion, we should remove GcNames completely and not do any of this pool-"interpreting" we do already.

I like the idea. I imagine it was added to make it easier to discuss the pools across different collectors.

What do you think about merging this as is and then opening up another issue/PR for removing GcNames and marking it discuss? I expect there are reasons it is there that I haven't thought of but I wonder if they outweigh problem it causes: it moves the vocabulary for "normal" optimizing java vocabulary to optimizing Elasticsearch vocabulary. And for lots of folks (me, probably @rmuir too) it is just a level of indirection that requires mental undo-ing when you want to think about the actual pools in use.

**Comment by rmuir** [comment]
OK I opened https://github.com/elastic/elasticsearch/issues/19720

**Comment by nik9000** [comment]
> OK I opened #19720


Cool. I still think you should merge this because it is an improvement even if we keep GcNames.

File: core/src/main/java/org/elasticsearch/monitor/jvm/JvmStats.java [hunk 1]
```diff
@@ -79,14 +79,15 @@ public static JvmStats jvmStats() {
...
+                    name = memoryPoolMXBean.getName();
                 }
+                // MemoryUsage.getMax() can be -1, indicating the maximum is undefined (e.g. metaspace max)
                 pools.add(new MemoryPool(name,
-                        usage.getUsed() < 0 ? 0 : usage.getUsed(),
-                        usage.getMax() < 0 ? 0 : usage.getMax(),
-                        peakUsage.getUsed() < 0 ? 0 : peakUsage.getUsed(),
-                        peakUsage.getMax() < 0 ? 0 : peakUsage.getMax()
+                        usage.getUsed(),
+                        usage.getMax() == -1 ? Long.MAX_VALUE : usage.getMax(),
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/monitor/jvm/JvmStats.java]: Why not just leave these at -1?

Comment (rmuir) [inline core/src/main/java/org/elasticsearch/monitor/jvm/JvmStats.java]: they are written with writeVLong which does not support negative numbers.

Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/monitor/jvm/JvmStats.java]: We can change that here and write them with `writeZLong`?

Comment (rmuir) [inline core/src/main/java/org/elasticsearch/monitor/jvm/JvmStats.java]: We can. It does mean people have to be prepared to understand -1 means unbounded, not unsupported, but in general I am for less "interpretation". I would just change them to writeLong and not play around. The numbers will be big and trying to compress them is overkill.

Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/monitor/jvm/JvmStats.java]: > in general I am for less "interpretation".

+1

> I would just change them to writeLong and not play around.

Sure, that's fine too.

Comment (rmuir) [inline core/src/main/java/org/elasticsearch/monitor/jvm/JvmStats.java]: I looked into this change, I don't think its reasonable to try to fold into this. The problem is that its hard to keep things as `-1` because `ByteSizeValue` is the api around all this, and its unclear that it supports negative values: there is no guard to prevent it, but at least its serialization would not work (uses vlong), who knows about its formatting and other stuff :)

Perhaps its possible to change the API to return null to indicate the unbounded case, but that's a lot of work if we want to apply it across the whole stats api too. Definitely requires simple serialization tests, which i started writing, but are missing everywhere, and there are bugs that reflect that, e.g. classloader statistics just get dropped on the floor.

Basically, I see a lot of bugs everywhere, and no clear path to fixing the situation without a lot of work. For now, I guess we can just remain in the dark on these statistics.
